### PR TITLE
Allow requests to match on POSTed parameters

### PR DESCRIPTION
**Pull request by samrussell**
Implements #313

### CODE REVIEW

**Comment by samrussell** [comment]
What's the deal with .5 and .6? The test runs seemed to be fine but a pre-commit task is failing?

File: responses.py [hunk 1]
```diff
@@ -312,6 +313,12 @@ def _url_matches(self, url, other, match_querystring=False):
         else:
             return False
 
+    def _post_params_match(self, post_params, request_body, content_type):
+        if post_params == None:
```
Comment (kathawala) [inline responses.py:317]: To satisfy the linter, this should be changed to

```python
if post_params is None:
```

Comment (samrussell) [reply responses.py:317]: nice catch, thank you! need more than a falsey comparison ({} would be valid to match on, for example)

Comment (samrussell) [reply responses.py:317]: what's a good way to run the linter locally? is it pylint or similar?

Comment (kathawala) [reply responses.py:317]: I think the linter used is [https://github.com/psf/black](url) if you want to run locally

File: responses.py [hunk 2]
```diff
@@ -312,6 +313,12 @@ def _url_matches(self, url, other, match_querystring=False):
         else:
             return False
 
+    def _post_params_match(self, post_params, request_body, content_type):
+        if post_params is None:
+            return True
+
+        return sorted(post_params.items()) == sorted(parse_qsl(request_body))
```
Comment (markstory) [inline responses.py:320]: How would this work for non URL encoded post bodies?

Comment (samrussell) [reply responses.py:320]: That's a very good point :) I have `x-www-form-urlencoded` in mind but obviously would need to check for that and exclude others.

The use case I'm looking at is testing a Slack integration to make sure it passes the right params upstream. It gets called with a list of users to invite, and I want to confirm that it's POSTing the correct IDs. The pattern looks something like this:

```
with responses.RequestsMock() as request_mock:
  request_mock.add(responses.POST, <matcher that matches the URL and POSTed params>, body=blah, status=200)
  slack_api.invite_users(["jonny", "mark", "lisa"])
```

My version only works with `x-www-form-urlencoded`, but then other payloads vary... Would it be better to allow responses/RequestsMock.add to take some sort of Matcher object that looks at the URL and then looks at the request headers/body to parse out params?

Comment (samrussell) [reply responses.py:320]: I've done a version that take custom matchers and added implementations for both JSON and urlencoding on #319. I'll have a go at a different version of post_params that uses the content-type to decide how to match

Comment (samrussell) [reply responses.py:320]: I got it working with the `post_params` argument, and there's code now that uses `Content-Type` to decide how to decode and compare the body - PR #320

Comment (samrussell) [reply responses.py:320]: Hi, have you had a chance to take another look at this? I think either #319 or #320 would both be good things that fit with the project, but if not I'm happy to try another tack

Comment (markstory) [reply responses.py:320]: I talked with some of the other folks at sentry and an approach similar to #319 is more along the lines of what we would prefer. Having a flexible single keyword argument that allows multiple matchers to be used is preferable over embedding implicit content-type logic or adding more parameters for each request dimension.

If its ok with you, I'd like to close this pull request and #320 and focus on #319

**Comment by samrussell** [comment]
That's fine by me, please let me know anything #319 is missing and I'm
happy to do some more work there

**Comment by RochesterinNYC** [comment]
Just found this feature/issue and noting that I/we would find the ability to do this in `responses` very useful as well.

**Comment by markstory** [comment]
Closing in favour of #319

**Pull request closed without merging**